lsp.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tmlsp
+
+import (
+	"context"
+	"encoding/json"
+	"log"
+
+	"go.lsp.dev/jsonrpc2"
+	lsp "go.lsp.dev/protocol"
+	"go.lsp.dev/uri"
+)
+
+type Server struct {
+	conn jsonrpc2.Conn
+
+	workspace string
+}
+
+type initParams struct {
+	ProcessID int    `json:"processId,omitempty"`
+	RootURI   string `json:"rootUri,omitempty"`
+}
+
+func NewServer(conn jsonrpc2.Conn) *Server {
+	return &Server{
+		conn: conn,
+	}
+}
+
+// Handler handles the client requests.
+func (s *Server) Handler(ctx context.Context, reply jsonrpc2.Replier, r jsonrpc2.Request) error {
+	log.Printf("got request %s with params: %s", r.Method(), r.Params())
```

Comment:
```suggestion
	log.Printf("got request %s with params: %s\n", r.Method(), r.Params())
```

Reply:
I would create a log function that does formatting  + always adds a newline, it is what I want 99% of the times but Go doesn't provide anything that does formatting + a newline (only Println), so it is very usual to forget the newlines and the log entries get clobbered.

Reply:
the `log` package already adds a newline. Why an extra one? :thinking:

Reply:
For some reason I thought it didn't, but yeah it actually does :laughing: (was probably thinking about fmt).

Reply:
![d](https://raw.githubusercontent.com/katcipis/memes/master/duh.jpeg)